### PR TITLE
Assignment6

### PR DESCRIPTION
**Pull request by jstanley6**
Project messed up working on fixing it

### CODE REVIEW

File: Assignment6/Calculate/Calculate/Program.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
+﻿// See https://aka.ms/new-console-template for more information
+
+namespace Calculate;
+
+using System;
+
+public class Program
+{
+    public Program()
```
Comment (Nolan-Posey) [inline Assignment6/Calculate/Calculate/Program.cs:9]: The Program constructor should be empty, I will leave a suggestion for how to set the default behavior for WrtieLine and ReadLine properties.

File: Assignment6/Calculate/Calculate/Program.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        Writeline = Console.WriteLine;
+        Readline = Console.ReadLine;
+    }
+
+    public Action<string> Writeline { get; init; }
```
Comment (Nolan-Posey) [inline Assignment6/Calculate/Calculate/Program.cs:15]: I recommend using a lambda expression like so:

public Action<string> WriteLine { get; init; } = text => Console.WriteLine(text);

File: Assignment6/Calculate/Calculate/Program.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    public Func<string?> Readline { get; init; }
```
Comment (Nolan-Posey) [inline Assignment6/Calculate/Calculate/Program.cs:16]: Same as above, I recommend using a lambda expression:

ReadLine { get; init; } = () = Console.ReadLine();

File: Assignment6/Calculate/Calculate/ProgramTests/CalculatorTests.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+    [TestMethod]
+    public void TestDictAreEqual()
+    {
+        Func<int, int, double> multiplication = Calculator.Multiply;
+        Func<int, int, double> division = Calculator.Divide;
+        Func<int, int, double> add = Calculator.Add;
+        Func<int, int, double> subtraction = Calculator.Subtract;
+        
+        Assert.AreEqual(multiplication, _calc.MathematicalOperators['*']);
```
Comment (Nolan-Posey) [inline Assignment6/Calculate/Calculate/ProgramTests/CalculatorTests.cs:18]: I recommend using the generics version of AreEqual for all four of these tests:

Assert.AreEqual( Func( int, int, double ) )

Just replace the parenthesis with angled brackets, I'm not sure why angled brackets don't show up.

File: Assignment6/Calculate/Calculate/ProgramTests/CalculatorTests.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        
+    }
+
+    [TestMethod]
+    public void Add_Test()
+    {
+        double addTest = Calculator.Add(6, 2);
+        double resTest = 8;
+        
+        Assert.AreEqual(resTest, addTest);
```
Comment (Nolan-Posey) [inline Assignment6/Calculate/Calculate/ProgramTests/CalculatorTests.cs:31]: Use generics version of AreEqual for a double.

File: Assignment6/Calculate/Calculate/ProgramTests/CalculatorTests.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    }
+    
+    [TestMethod]
+    public void Subtract_Test()
+    {
+        double subTest = Calculator.Subtract(6, 2);
+        double resTest = 4;
+        
+        Assert.AreEqual(resTest, subTest);
```
Comment (Nolan-Posey) [inline Assignment6/Calculate/Calculate/ProgramTests/CalculatorTests.cs:40]: Use generics version of AreEqual for a double.

File: Assignment6/Calculate/Calculate/ProgramTests/CalculatorTests.cs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    }
+    
+    [TestMethod]
+    public void Multiply_Test()
+    {
+        double multiplyTest = Calculator.Multiply(8, 2);
+        double resTest = 16;
+        
+        Assert.AreEqual(resTest, multiplyTest);
```
Comment (Nolan-Posey) [inline Assignment6/Calculate/Calculate/ProgramTests/CalculatorTests.cs:49]: Use generics version of AreEqual for a double.

File: Assignment6/Calculate/Calculate/ProgramTests/CalculatorTests.cs [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    }
+    
+    [TestMethod]
+    public void Divide_Test()
+    {
+        double divTest = Calculator.Divide(10, 5);
+        double resTest = 2;
+        
+        Assert.AreEqual(resTest, divTest);
```
Comment (Nolan-Posey) [inline Assignment6/Calculate/Calculate/ProgramTests/CalculatorTests.cs:58]: Use generics version of AreEqual for a double.

File: Assignment6/Calculate/Calculate/ProgramTests/ProgramTests.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    [TestMethod]
+    public void TestWriteLine()
+    {
+        StringWriter stringWriter = new();
+        Console.SetOut(stringWriter);
+        string expected = "expected";
+
+        new Program().Writeline(expected);
+        
+        Assert.AreEqual(expected, stringWriter.ToString().Trim());
```
Comment (Nolan-Posey) [inline Assignment6/Calculate/Calculate/ProgramTests/ProgramTests.cs:18]: Use generics version of AreEqual for a string.

File: Assignment6/Calculate/Calculate/ProgramTests/ProgramTests.cs [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+        StringReader stringReader = new(userInput);
+        Console.SetIn(stringReader);
+        string expected = "The result of the equation 10 * 10 is 100";
+        userInput = program.Readline();
+        
+        Assert.IsTrue(calculator.TryCalculate(userInput!, out double result));
+        program.Writeline($"The result of the equation {userInput} is {result}");
+        
+        Assert.AreEqual(expected, stringWriter.ToString().Trim());
```
Comment (Nolan-Posey) [inline Assignment6/Calculate/Calculate/ProgramTests/ProgramTests.cs:49]: Use generics version of AreEqual for a string.

**Review by Nolan-Posey — COMMENTED** [review]
Create a Console project called "Calculate.". ✔
Define a Program Class
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ❌
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ✔
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ✔
Extra Credit
Do one of the following two options (or both if you want extra, extra credit) :)

Refactor the redirect portion of the Program class into 'ProgramBase`
Move ProgramBase into a ConsoleUtilities assembly to be used in other console-based projects
Use generics the mathematical operations methods and consider using generic constraints (requires .NET 7.0)
Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ❌
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

Overall you did a great job! It seems you were able to resolve the issue you were having with the Directory.Build.props file. You just missed a couple things but I added some comments offering my suggestions to fix those problems.  I have some other suggestions to add that might help you on future projects:

1.) Add a test project to your solution to help organize your tests:

If you right click on the project solution while in the solution explorer, you can add a new project. Click to add a new project and select "MSTest Test Project". If you can't find it you will have to click the "Online" tab on the left and download it from there. Once you have it and make a new MSTest Test Project you can name it "CalculateTests" for example, and then you can add all your test classes to it. This will be a lot easier than having to make a folder in the solution and make organizing your tests much more simple.

2.) How I add a Directory.Build.props file to my projects (recommended for future projects):

When I make my Directory.Build.props file for my projects I begin by right clicking on the solution file from the solution explorer. From there I click Add, then click New Item. From there, delete everything in the "Name" bar at the bottom of the pop-up screen and type "Directory.Build.props" then click add. from there Visual Studio should make a "Solution Items" folder on the right level of the directory hierarchy and add the Directory.Build.props file to the folder. Then all you have to do is add the shard properties from both your main project and your test project to the file. For a "Directory.Build.targets" file  you can repeat the same process as before but just add the "New Item" to the "Solution Items" folder that was created for you.

Hope this helps!
-Nolan Posey

File: Assignment6/Calculate/Calculate/ProgramTests/CalculatorTests.cs (hunk 4 above)
Comment (jstanley6) [reply Assignment6/Calculate/Calculate/ProgramTests/CalculatorTests.cs:18]: I did the generic version at first but it gave me a warning to remove it and it wasn't needed.

File: Assignment6/Calculate/Calculate/Program.cs (hunk 1 above)
Comment (jstanley6) [reply Assignment6/Calculate/Calculate/Program.cs:9]: done

File: Assignment6/Calculate/Calculate/Calculator.cs [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+    public static double Multiply(int a, int b)
+        => (double)a * b;
+
+    public static double Divide(int a, int b)
+        => (double)a / b;
+
+    public bool TryCalculate(string calculation, out double result)
+    {
+        var split = calculation.Split(" ");
```
Comment (mmwoodfo) [inline Assignment6/Calculate/Calculate/Calculator.cs:36]: You should be using the Split method along side pattern matching. e.g.
`if (expression?.Split(" ") is [string operator1, [char mathmaticalOperator], string operand2])`

**Pull request closed without merging**